e2e/controls.py

```diff
@@ -0,0 +1,116 @@
+"""Facade classes for working with Shiny inputs/outputs in Playwright"""
+
+# pyright: reportUnknownMemberType=false
+
+from playwright.sync_api import Page, expect, Locator
+
+
+class SimpleInput:
+    def __init__(self, page: Page, selector: str):
+        self.loc = page.locator(selector)
+
+    @property
+    def expect(self):
+        return expect(self.loc)
+
+
+class TextInput(SimpleInput):
+    def __init__(self, page: Page, inputId: str):
+        super().__init__(page, f"input#{inputId}[type=text].shiny-bound-input")
+
+
+class TextAreaInput(SimpleInput):
+    def __init__(self, page: Page, inputId: str):
+        super().__init__(page, f"textarea#{inputId}.shiny-bound-input")
+
+
+class NumericInput(SimpleInput):
+    def __init__(self, page: Page, inputId: str):
+        super().__init__(page, f"input#{inputId}[type=number].shiny-bound-input")
+
+
+class CheckboxInput(SimpleInput):
+    def __init__(self, page: Page, inputId: str):
+        super().__init__(page, f"input#{inputId}[type=checkbox].shiny-bound-input")
+
+
+class DateInput:
+    def __init__(self, page: Page, inputId: str):
+        self.loc = page.locator(f"#{inputId}")
+
+    @property
+    def expect(self):
+        return expect(self.loc.locator("input"))
+
+
+class DateRangeInput:
+    def __init__(self, page: Page, inputId: str):
+        self.loc = page.locator(f"#{inputId}")
+
+    @property
+    def expect_start(self):
+        return expect(self.loc.locator("input:first-of-type"))
+
+    @property
+    def expect_end(self):
+        return expect(self.loc.locator("input:last-of-type"))
+
+
+class SliderInput:
```

Comment:
How come there is no `expect` property here? Is it because the markup is different  depending on whether `value` is of length 1 or 2?

Reply:
I was just creating those on-demand. Especially because I just didn't think about it very much as I was creating controls.py. Barret is going to do a more thoughtful buildout of controls.py.